New "enabled" meta-attribute for alerts

nrwahl2 opened the pull request:
An `<alert>` element or an individual `<recipient>` element can now contain a boolean `enabled` meta attribute. The default value is "true". If `enabled` is set to "false", the alert or recipient will not be used.

If `enabled` is set to "false" at the `<alert>` level, all of the alert's recipients are disabled, regardless of the recipients' `enabled` value.

If `enabled` is set to "true" or is unset at the `<alert>` level, the alert's recipients are enabled or disabled based on their respective values of the `enabled` meta attribute.

Setting `enabled` to "false" can be useful to avoid spamming alert recipients during planned maintenance, for example.

nrwahl2 commented on lib/pengine/rules_alerts.c:
```diff
@@ -27,6 +27,13 @@ get_meta_attrs_from_cib(xmlNode *basenode, pcmk__alert_t *entry,
+    value = g_hash_table_lookup(config_hash, PCMK_XA_ALERT_ENABLED);
+    if ((value != NULL) && !crm_is_true(value)) {
+        // No need to continue unpacking
+        entry->enabled = false;
```
This is my preferred approach.

We *could* avoid adding the `pcmk__alert_t:enabled` member and instead set the `entry->flags` to `pcmk__alert_none` if `enabled=false`. However, we would do that in `unpack_alert_filter()`, which requires that we finish unpacking the alert that we're not going to use.

Another option would be to set `entry->flags = pcmk__alert_none` here, and check that before calling `unpack_alert_filter()`. That would work since the default value of `entry->flags` is not `pcmk__alert_none` (it's nodes|fencing|resources). But that feels less clear to me since `unpack_alert_filter()` is what's really supposed to set those flags.

kgaillot replied:
With your implementation, we don't really need to track disabled in the object at all, we just need get_meta_attrs_from_cib() and unpack_alert() to return success/skip

nrwahl2 replied:
Oh you're right! When I started working on this and added that member, I thought it was going to be a little bit more complicated than it turned out to be.

kgaillot commented on doc/sphinx/Pacemaker_Explained/alerts.rst:
```diff
@@ -92,6 +92,16 @@ to affect how Pacemaker calls them.
+   |                  |               | If false for an alert agent, the alert agent will   |
+   |                  |               | not be used. If true for an alert agent and false   |
+   |                  |               | for a particular recipient of that agent, that      |
+   |                  |               | recipient will not be used.                         |
```
I would say just "alert" instead of "alert agent" (you could have multiple alerts with the same agent for various reasons)

nrwahl2 replied:
I totally agree. I was trying to be consistent with the rest of this section, which says "alert agents" when it means alerts. I can update the rest where appropriate.

nrwahl2 replied:
updated

kgaillot reviewed (commented):
Always nice when there's an easy win :)

nrwahl2 commented on include/crm/msg_xml.h:
```diff
@@ -402,6 +402,7 @@ extern "C" {
 #  define XML_CONFIG_ATTR_SHUTDOWN_LOCK_LIMIT   "shutdown-lock-limit"
 #  define XML_CONFIG_ATTR_PRIORITY_FENCING_DELAY "priority-fencing-delay"
 
+#  define PCMK_XA_ALERT_ENABLED		"enabled"
```
Makes sense

nrwahl2 replied:
makes me wonder if constants used for meta attributes should continue to be public going forward, or internal like `PCMK__VALUE_*`

nrwahl2 replied:
Updated to use `PCMK_META_ENABLED`. I'm still open to `PCMK_NAME_*`, and I think that makes pretty good sense if we make these internal constants. I'm less sure about having `PCMK_NAME_*` as public and `PCMK__VALUE` as private.

kgaillot replied:
Yeah I made the VALUE constants private just to give us some time before having to decide whether they should be or not. There are arguments on both sides. I think it does make sense that external code might want the safety of a constant when manipulating CIBs or such. I think META and public are reasonable choices.

nrwahl2 commented:
* Added new `pcmk_rc_disabled` to use with `get_meta_attrs_from_cib()`.
* Changed string constant name to `PCMK_META_ENABLED`.
* Removed `pcmk__alert_t:enabled` and used return code instead
* Updated doc to say "alert" instead of "alert agent" in the Alert Meta-attributes section. Some of the other uses throughout the Alerts section are a grayer area. I left them alone for now.

kgaillot commented on lib/common/results.c:
```diff
@@ -293,6 +293,10 @@ static const struct pcmk__rc_info {
       -pcmk_err_generic,
     },
+    { "pcmk_rc_disabled",
+      "Object is disabled",
```
I'm thinking just "Disabled". The idea is that the log message has more context than "object" like `"Ignoring alert %s: %s", alert_id, pcmk_rc_str(rc)`

nrwahl2 replied:
I prefer that too. I was afraid you'd say it was too short haha. The current version offers no additional info though.

kgaillot commented on include/crm/msg_xml.h:
```diff
@@ -36,8 +36,11 @@ extern "C" {
 
+#  define PCMK_XE_DATE_EXPRESSION           "date_expression"
+#  define PCMK_XE_OP_EXPRESSION             "op_expression"
+#  define PCMK_XE_RSC_EXPRESSION            "rsc_expression"
```
I've never been thrilled with indenting defines; I think it takes up space and looks weird with more than one level of indenting, without particularly helping readability. Also these should in alphabetic order for ease of maintenance once everything is moved here.

nrwahl2 replied:
I've never indented defines in my own code unless it's maybe in an immediate block.

I was keeping them in "logical" blocks (arbitrary) but I'm happy with alphabetical. Eliminates the need to spend time deciding "where does this new one go?"

kgaillot reviewed (commented):
Looks great
